arangod/Graph/Enumerators/WeightedTwoSidedEnumerator.h

```diff
@@ -59,18 +63,96 @@ template<class QueueType, class PathStoreType, class ProviderType,
 class WeightedTwoSidedEnumerator {
  public:
   using Step = typename ProviderType::Step;  // public due to tracer access
-  using Candidate = std::pair<Step, Step>;
-  using CandidatesMap = std::unordered_map<double, Candidate>;
+
+  // A meeting point with calculated path weight
+  using CalculatedCandidate = std::tuple<double, Step, Step>;
 
  private:
   enum Direction { FORWARD, BACKWARD };
 
   using VertexRef = arangodb::velocypack::HashedStringRef;
 
   using Shell = std::multiset<Step>;
-  using ResultList = std::vector<std::pair<Step, Step>>;
+  using ResultList = std::deque<CalculatedCandidate>;
   using GraphOptions = arangodb::graph::TwoSidedEnumeratorOptions;
 
+  /*
+   * Weighted candidate store for handling path-match candidates in the order
+   * from the lowest weight to the highest weight.
+   */
+  class CandidatesStore {
+   public:
+    void clear() {
+      if (!_queue.empty()) {
+        _queue.clear();
+      }
+    }
+
+    void append(CalculatedCandidate candidate) {
+      // if emplace() throws, no harm is done, and the memory usage increase
+      // will be rolled back
+      _queue.emplace_back(std::move(candidate));
+      // std::push_heap takes the last element in the queue, assumes that all
+      // other elements are in heap structure, and moves the last element into
+      // the correct position in the heap (incl. rebalancing of other elements)
+      // The heap structure guarantees that the first element in the queue
+      // is the "largest" element (in our case it is the smallest, as we
+      // inverted the comparator)
+      std::push_heap(_queue.begin(), _queue.end(), _cmpHeap);
+    }
+
+    size_t size() const { return _queue.size(); }
+
+    bool isEmpty() const { return _queue.empty(); }
+
+    std::vector<CalculatedCandidate> getQueue() const { return _queue; };
+
+    CalculatedCandidate& peek() {
+      TRI_ASSERT(!_queue.empty());
+      // will return a pointer to the step with the lowest weight amount
+      // possible. The heap structure guarantees that the first element in the
+      // queue is the "largest" element (in our case it is the smallest, as we
+      // inverted the comparator)
+      auto& first = _queue.front();
+      return first;
+    }
+
+    CalculatedCandidate pop() {
+      TRI_ASSERT(!isEmpty());
+      // std::pop_heap will move the front element (the one we would like to
+      // steal) to the back of the vector, keeping the tree intact otherwise.
+      // Now we steal the last element.
+      std::pop_heap(_queue.begin(), _queue.end(), _cmpHeap);
+      CalculatedCandidate first = std::move(_queue.back());
+      _queue.pop_back();
+      return first;
+    }
+
+   private:
+    struct WeightedComparator {
+      bool operator()(CalculatedCandidate const& a,
+                      CalculatedCandidate const& b) {
+        auto const& [weightA, candAA, candAB] = a;
+        auto const& [weightB, candBA, candBB] = b;
+        if (weightA == weightB) {
```

Comment:
Comparing floating point numbers for equality is likely the wrong thing to do.

Reply:
removed.